### ISSUE TITLE
Support for named parameters is broken in Python 3

### ISSUE THREAD

**Issue by FredrikAppelros** [uber/vertica-python#112]
Support for named parameters is not working in Python 3. This seems to stem from how psycopg2 is used to quote the parameters.

The following example demonstrates how it breaks down for certain unicode characters:

```
>>> from vertica_python import connect
>>> conn = connect(host='localhost', database='docker', user='dbadmin', password='')
>>> cursor = conn.cursor()
>>> cursor.execute('SELECT :param', {'param': u'\u16b1'})
>>> cursor.fetchone()
[b'\\341\\232\\261']
```

We can also see it breaking down for ordinary characters that should not give us any issue regardless of what encoding we are using:

```
>>> from vertica_python import connect
>>> conn = connect(host='localhost', database='docker', user='dbadmin', password='')
>>> cursor = conn.cursor()
>>> cursor.execute('CREATE TABLE test (test_col VARCHAR)')
>>> cursor.execute('SELECT 1 FROM test WHERE test_col = :param', {'param': 'test123'})
Traceback (most recent call last):
  File "<input>", line 1, in <module>
    cursor.execute('SELECT 1 FROM test WHERE test_col = :param', {'param': 'test123'})
  File "/home/fredrik/projects/vertica-python/vertica_python/vertica/cursor.py", line 93, in execute
    raise errors.QueryError.from_error_response(message, operation)
vertica_python.errors.QueryError: Severity: b'ERROR', Message: b'Operator does not exist: varchar = varbinary', Sqlstate: b'42883', Hint: b'No operator matches the given name and argument type(s). You may need to add explicit type casts', Routine: b'op_error', File: b'parse_oper.c', Line: b'577', SQL: "SELECT 1 FROM test WHERE test_col = 'test123'::bytea"
```

The issue comes from [this line](https://github.com/uber/vertica-python/blob/master/vertica_python/vertica/cursor.py#L64). It results in a byte array that represents the quoted original string, but with `::bytea` attached to the end. This in turn causes both of the issues I listed above. See the following example of what `adapt(...).getquoted()` does with UTF-8 encoded byte arrays.

```
>>> from psycopg2.extensions import adapt
>>> adapt('test123'.encode('utf8')).getquoted()
b"'test123'::bytea"
```

**Comment by FredrikAppelros** [comment]
Actually, in the first case, it's not just the extra `::bytea` that's causing the issue. It incorrectly quotes the backslashes in the byte array, causing them to be interpreted literally, instead of as escape characters.

	>>> from psycopg2.extensions import adapt
	>>> encoded = '\u16b1'.encode('utf8')
	>>> adapt(encoded).getquoted()
	b"'\\\\341\\\\232\\\\261'::bytea"
	>>> encoded
	b'\xe1\x9a\xb1'

**Comment by FredrikAppelros** [comment]
Not sure what the proper solution is here. Either we need to configure psycopg2 to quote the byte arrays the way we want it to, or we might need to write a new implementation for the quotation logic and remove the psycopg2 dependency.

Seems like psycopg2 can be configured to perform a proper quotation if the encoding is set to UTF-8:

	>>> from psycopg2 import connect
	>>> from psycopg2.extensions import adapt
	>>> conn = connect(host='localhost', user='postgres', password='secret', port=15432)
	>>> conn.set_client_encoding('utf8')
	>>> adpt = adapt('\u16b1')
	>>> adpt.prepare(conn)
	>>> adpt.getquoted()
	b"'\xe1\x9a\xb1'"

Now the problem is that I don't believe it's possible to change the encoding without creating an actual connection first, which of course is out of the question. I think that psycopg/psycopg2#331 might be relevant.

**Comment by FredrikAppelros** [comment]
I improved one of the existing testcases to capture this issue in #113.

**Comment by alexjikim** [comment]
thanks for the investigation into this. the psycopg2 dependency is one that id rather not have, but implementing custom logic for quoting/escaping seemed very error prone.

when i get a chance ill try to poke around the psycopg2 code and see if theres a workaround.

**Comment by veritas718** [comment]
Here is a code workaround that I dont love, but correctly substitutes most variable names:
    ```
def vsql(self, sql, params):
        st = time.time()
        with vertica_python.connect(**self.conn_info) as connection:
            cur = connection.cursor()
            for parameter in params:
                sql=sql.replace((':'+parameter),str(params[parameter]))
            print(sql)
            cur.execute(sql)
            connection.commit()
        print("{0:.3f} s \t {1}".format(time.time()-st, sql))
        return
```

**Comment by twneale** [comment]
This one has an easy fix that doesn't require a big decision about psycopg2: see #129 for details

**Comment by alexjikim** [comment]
reopening: still broken for python3 unicode case

**Comment by jbfavre** [comment]
Hello,
I'm the maintainer for python-vertica package in Debian.
I'd like to let you know that current Debian testing will be freezed on 5th of february 2017 (https://wiki.debian.org/DebianStretch)

I'd really like to see this issue can be fixed before the freeze so that I can upload a fully functional package into Debian archive.
If it's fixed after 2017-02-05, the fix won't be available in Stretch.

Regards,
JB

**Comment by feinsteinben** [comment]
I think that #152 solves it (there are many unicode/compatiblity fixes there). I ran all tests on py27, py34, py35, py36 and it worked well (including the usage of named parameters).

**Comment by alexjikim** [comment]
this _seems_ to be resolved as of 0.7.0 with the PR by @feinsteinben
